### ISSUE TITLE
IDE Assist for `&str` -> `String`

### ISSUE THREAD

**Issue by gleich** [rust-analyzer/rust-analyzer#11383]
Hello, rust analyzer maintainers! It would be awesome to have an IDE assist feature in which a user could convert a `&str` to a `String` by essentially just wrapping it in `String::from()` (or `.to_string()`, whatever would be best). I would be interested in adding myself if it ends up getting accepted by the RA team.

**Comment by Veykril** [comment]
An assist for a trivial(and very specific task) like this seems unlikely to be of much value. This sounds more like the task for a custom completion snippet:
```json
    "rust-analyzer.completion.snippets": {
        "String::from": {
            "postfix": "strfrom",
            "body": "String::from(${receiver})",
            "scope": "expr",
        },
    },
```

On the other hand, we should already show a `to_string` completion for you when putting a `.` after a `&str` variable.

**Comment by flodiebold** [comment]
> An assist for a trivial(and very specific task) like this seems unlikely to be of much value

Slightly disagree, once we have a diagnostic saying "this expects a String, but got a &str", then it makes sense to give a quickfix for that.

**Comment by Veykril** [comment]
Yes of course, I was solely thinking of assists, not diagnostic quickfixes. We should definitely have a diagnostic here.

**Comment by flodiebold** [comment]
This is doable now that we have the `type-mismatch` diagnostic.
Quickfixes are defined here:
https://github.com/rust-analyzer/rust-analyzer/blob/9da9418661c2a357b291d6adf11ea296d9dd924a/crates/ide_diagnostics/src/handlers/type_mismatch.rs#L32-L44

**Comment by harpsword** [comment]
https://user-images.githubusercontent.com/10148927/165762695-00b60388-e153-4b5b-bff6-ac141e3a35e7.mov

I found out that this function already exists.
